src/proposed/0180-fpu-switching.md

```diff
@@ -0,0 +1,378 @@
+<!--
+  SPDX-License-Identifier: CC-BY-SA-4.0
+  Copyright 2024 Proofcraft Pty Ltd
+-->
+
+# FPU Context Switching
+
+- Author: Gerwin Klein, Rafal Kolanski, Indan Zupancic
+- Proposed: 2024-06-21
+
+## Summary
+
+We propose to let user level configure policy on when FPU context is switched
+instead of relying on runtime faults to switch FPU context lazily. Policy is
+configured via a flag in the TCB, authorised by the corresponding TCB
+capability.
+
+This improves performance, moves policy from kernel to user, enables verification
+of FPU handling, and fixes an information leakage in current FPU context
+switching.
+
+## Motivation
+
+Current FPU context switching in seL4 disables the FPU on kernel exit, waits for
+a thread to fault when it accesses the FPU, transparently handles the fault,
+saves/restores FPU context, and then returns to the thread.
+
+The motivation for this scheme was that threads that never use the FPU do not pay
+for FPU context switching cost. However, it has the following problems:
+
+- this is a fixed context switching policy in the kernel;
+
+- most of this complexity is invisible to the current formal verification, and
+  therefore not covered by the proofs;
+
+- when switching between domains, this scheme introduces information flow
+  leakage, because the kernel performs actions on behalf of domain A while
+  running in domain B: assuming A and B are the only domains, the first thread
+  running in domain B can determine whether a thread in domain A used the FPU
+  simply by using the FPU itself, since kernel FPU save/restore actions will
+  only occur if another thread used the FPU.
+
+- performance of this scheme is not optimal -- taking a fault every time the FPU
+  is used adds unnecessary overhead, especially in systems where FPU usage is
+  high.
```

Comment:
This and the line at the start of this section ("Current FPU context switching in seL4 disables the FPU on kernel exit,") could be misleading. The kernel doesn't disable the FPU on every kernel exit.  Just when the current thread is not the current owner of the FPU. If the current thread is also the current owner of the FPU, then the FPU is enabled.

Reply:
Agreed it should be made clearer that faulting at every FPU usage only happens when multiple threads use the FPU. On x86 this seems to be happening now because compilers are eager to use SIMD instructions even for integer operations, so in practise you get one FPU fault every context switch.

Reply:
Have tweaked the text a bit to clarify that.